### PR TITLE
WIP Allow updating the cluster one instance group at a time

### PR DESCRIPTION
**Pull request by hakman**
/cc @justinsb @rifelpet

### CODE REVIEW

File: cmd/kops/update_cluster.go [hunk 1]
```diff
@@ -285,24 +306,75 @@ func RunUpdateCluster(ctx context.Context, f *util.Factory, out io.Writer, c *Up
 		lifecycleOverrideMap[taskName] = lifecycleOverride
 	}
 
+	list, err := clientset.InstanceGroupsFor(cluster).List(ctx, metav1.ListOptions{})
```
Comment (justinsb) [inline cmd/kops/update_cluster.go:309]: Nit: might be nice to split this into a function?

**Comment by justinsb** [comment]
Remind me - I think you said there's a gotcha with this approach @hakman ?

**Comment by hakman** [comment]
> Remind me - I think you said there's a gotcha with this approach @hakman ?

@justinsb If you try to apply the config to a single IG, the KubeletConfig is removed.

**Comment by justinsb** [comment]
Looked at this for a while, and I think I know what is happening!

So the issue is:

```
// upgradeSpecs ensures that fields are fully populated / defaulted
func (c *ApplyClusterCmd) upgradeSpecs(ctx context.Context, assetBuilder *assets.AssetBuilder) error {
```

We store the instanceGroup as the user specified it.  But there are many computed fields - the so called "full" instanceGroup or cluster spec.  upgradeSpecs converts the instanceGroups to populate these fields.

In this PR, we aren't calling upgradeSpecs or the population logic.  So these instance groups don't have many fields populated, including (I would guess) the kubelet config.

I suggest that the easy thing to do is to convert `FilteredInstanceGroups` into a predicate (i.e. `func() bool`), and then basically filter these out at the last minute.  Then we don't need to duplicate the defaulting logic.

**Comment by justinsb** [comment]
Applying a predicate does seem to work ... https://github.com/justinsb/kops/tree/update-cluster-per-ig

The problem is that I don't think we can just do ASGs.  Even on AWS, we need to deal with NodeTerminationHandler object.  But I think this problem also exists on other clouds, so we need to solve it there also IIUC

**Comment by justinsb** [comment]
I opened another PR #16936 with the filter.  I also had to do something similar to what you did - _all_ tasks have access to InstanceGroups (the filtered instance groups) and AllInstanceGroups (unfiltered).  The "trick" is that most tasks work with "InstanceGroups" (the filtered), there are just a handful that need to be repointed to work with AllInstanceGroups (e.g. things that need a list of roles)

Likely we'll need a few tweaks for other clouds, but they shouldn't be too big!

**Comment by hakman** [comment]
> I opened another PR #16936 with the filter. I also had to do something similar to what you did - _all_ tasks have access to InstanceGroups (the filtered instance groups) and AllInstanceGroups (unfiltered). The "trick" is that most tasks work with "InstanceGroups" (the filtered), there are just a handful that need to be repointed to work with AllInstanceGroups (e.g. things that need a list of roles)
> 
> Likely we'll need a few tweaks for other clouds, but they shouldn't be too big!

Awesome! 🎉

**Pull request closed without merging**